src/v/cluster/tests/feature_barrier_test.cc

```diff
@@ -0,0 +1,259 @@
+ */
+FIXTURE_TEST(test_barrier_simple, barrier_fixture) {
+    create_brokers(3);
+    create_node_state(model::node_id{0});
+    create_node_state(model::node_id{1});
+    create_node_state(model::node_id{2});
+
+    auto f0 = get_barrier_state(model::node_id{0})
+                .barrier(feature_barrier_tag{"test"});
+    auto f1 = get_barrier_state(model::node_id{1})
+                .barrier(feature_barrier_tag{"test"});
+
+    BOOST_REQUIRE(!f0.available());
+    BOOST_REQUIRE(!f1.available());
+
+    auto f2 = get_barrier_state(model::node_id{2})
+                .barrier(feature_barrier_tag{"test"});
+
+    ss::sleep(10ms).get();
+
+    BOOST_REQUIRE(f0.available());
+    BOOST_REQUIRE(f1.available());
+    BOOST_REQUIRE(f2.available());
+
+    // Now that all three have started running, all should complete
+    f0.get();
+    f1.get();
+    f2.get();
+}
+
+FIXTURE_TEST(test_barrier_node_restart, barrier_fixture) {
+    create_brokers(3);
+    create_node_state(model::node_id{0});
+    create_node_state(model::node_id{1});
+    create_node_state(model::node_id{2});
+
+    // Nodes 1+2 enter the barrier
+    auto f0 = get_barrier_state(model::node_id{0})
+                .barrier(feature_barrier_tag{"test"});
+    auto f1 = get_barrier_state(model::node_id{1})
+                .barrier(feature_barrier_tag{"test"});
+
+    BOOST_REQUIRE(!f0.available());
+    BOOST_REQUIRE(!f1.available());
+
+    // Node 3 enters the barrier
+    auto f2 = get_barrier_state(model::node_id{2})
+                .barrier(feature_barrier_tag{"test"});
+
+    // Prompt reactor to process outstanding futures
+    ss::sleep(10ms).get();
+
+    BOOST_REQUIRE(f0.available());
+    BOOST_REQUIRE(f1.available());
+
+    // Prompt reactor to process outstanding futures
+    ss::sleep(10ms).get();
+
+    restart(model::node_id{2});
+    f2 = get_barrier_state(model::node_id{2})
+           .barrier(feature_barrier_tag{"test"});
+
+    // Prompt reactor to process outstanding futures
+    ss::sleep(10ms).get();
+
+    BOOST_REQUIRE(f2.available());
+
+    // All futures are ready.
+    f0.get();
+    f1.get();
+    f2.get();
+}
+
+FIXTURE_TEST(test_barrier_node_isolated, barrier_fixture) {
+    create_brokers(3);
+    create_node_state(model::node_id{0});
+    create_node_state(model::node_id{1});
+    create_node_state(model::node_id{2});
+
+    rpc_rx_errors[model::node_id{2}] = cluster::make_error_code(
+      cluster::errc::timeout);
+    rpc_tx_errors[model::node_id{2}] = cluster::make_error_code(
+      cluster::errc::timeout);
+
+    auto f0 = get_barrier_state(model::node_id{0})
+                .barrier(feature_barrier_tag{"test"});
+    auto f1 = get_barrier_state(model::node_id{1})
+                .barrier(feature_barrier_tag{"test"});
+    auto f2 = get_barrier_state(model::node_id{2})
+                .barrier(feature_barrier_tag{"test"});
+
+    // Without comms to node 2, this barrier should block to complete
+    ss::sleep(1000ms).get();
```

Comment:
Intending to come back later and use `manual_clock` to make this cleaner.

Reply:
(this is done in https://github.com/redpanda-data/redpanda/pull/3982 but that'll land separately to avoid disrupting this PR)